Add CMake for onnx app

vawale opened the pull request:
Adds CMakeLists.txt file for cmake build of onnx app. The build logic is kept same as the one from corresponding Makefile.

Fixes [#8737](https://github.com/halide/Halide/issues/8737)

alexreinking commented on apps/onnx/CMakeLists.txt:
```diff
@@ -0,0 +1,162 @@
+find_package(Halide REQUIRED)
+
+# Download onnx.proto
+set(ONNX_PROTO_FILE ${CMAKE_CURRENT_BINARY_DIR}/onnx.proto)
+if (NOT EXISTS ${ONNX_PROTO_FILE})
+    file(
+        DOWNLOAD
+        "https://raw.githubusercontent.com/onnx/onnx/v1.4.1/onnx/onnx.proto"
+        ${ONNX_PROTO_FILE}
+        SHOW_PROGRESS
+        STATUS download_status
+    )
+    if (NOT download_status EQUAL 0)
+        message(FATAL_ERROR "Failed to download onnx.proto: ${download_status}")
+    endif ()
+
+    file(READ ${ONNX_PROTO_FILE} onnx_proto_content)
+    string(REPLACE "package onnx;" "package onnx;option optimize_for = LITE_RUNTIME;" onnx_proto_content "${onnx_proto_content}")
+    file(WRITE ${ONNX_PROTO_FILE} "${onnx_proto_content}")
+endif ()
```
I'd like to see this replaced by a call to `FetchContent`. It should be terser that way and it will support using a Git hash as a reference so we can be somewhat more sure the file hasn't been changed on the remote.

alexreinking commented on apps/onnx/CMakeLists.txt:
```diff
@@ -0,0 +1,162 @@
+)
+
+# Add library that converts ONNX models to Halide operators
+add_library(oclib STATIC onnx_converter.cc ${CMAKE_CURRENT_BINARY_DIR}/onnx/onnx.pb.cc)
+target_include_directories(
+    oclib
+    PUBLIC
+    ${CMAKE_CURRENT_BINARY_DIR}
+    ${Protobuf_INCLUDE_DIRS}
+)
```
We can probably simplify all this and the custom target that drives protobuf by using `FILE_SETS`.

alexreinking commented on apps/onnx/CMakeLists.txt:
```diff
@@ -0,0 +1,162 @@
+
+pybind11_add_module(model_cpp model.cpp)
+target_link_libraries(model_cpp PRIVATE Halide::Halide oclib)
+if (UNIX)
+    target_link_libraries(model_cpp PRIVATE -Wl,--no-as-needed Halide::Adams2019 -Wl,--as-needed)
+endif ()
```
I doubt this will work with the Apple linker... in `Makefile.inc`, @abadams uses `-force_load`, but I doubt that will work for a loadable module (`.so`) these days. Does the Makefile build the autoschedulers as static libraries ever?

alexreinking commented on apps/onnx/CMakeLists.txt:
```diff
@@ -0,0 +1,162 @@
+    MAIN_DEPENDENCY ${CMAKE_CURRENT_SOURCE_DIR}/test_model_proto.txt
+    VERBATIM
+)
+add_custom_target(test_model_onnx DEPENDS test_model.onnx)
```
```suggestion
add_custom_target(test_model_onnx DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/test_model.onnx")
```

Custom target dependencies are relative to the source dir, so the absolute path is needed here.

alexreinking reviewed (changes requested):
Still waiting on buildbots, but this is what stood out to me on a first pass.

vawale commented:
Thanks for the review and the follow up fixes. I will look into addressing the comments soon :)

> Some of the buildbots were missing protoc. Now I see the following errors:

Ah okay. Should I skip building this app if `protoc` is not available?

alexreinking commented:
New failures on the mac bots because WebGPU is missing. This is because they're running under launchd now and don't have the environment variables set. This is brittle, so I'm going to do this The Right Way...

https://github.com/Homebrew/homebrew-core/pull/232863

alexreinking commented:
> This is brittle, so I'm going to do this The Right Way...

Turns out this is too hard. See #8714 .

alexreinking commented:
@vawale -- I see that the halide_as_onnx_backend_test is timing out (and that subtests are failing). Does it work for you?

https://buildbot.halide-lang.org/master/#/builders/265/builds/30

vawale commented:
> Seems like `halide_as_onnx_backend_test` is failing because Adams2019 is thinking for too long.

Yes, I observe the same. The auto-scheduler takes too long.
